sonar-home/src/main/java/org/sonar/home/cache/PersistentCache.java

```diff
@@ -0,0 +1,266 @@
+      }
+      else {
+        log.debug("cache force update for " + obj);
+      }
+
+      if (valueLoader != null) {
+        byte[] value = valueLoader.call();
+        if (value != null) {
+          putCache(key, value);
+        }
+        return value;
+      }
+    }
+
+    return null;
+  }
+
+  /**
+   * Deletes all cache entries
+   */
+  public synchronized void clear() {
+    log.info("cache: clearing");
+    try (FileLock l = lock()) {
+      deleteCacheEntries(createClearFilter());
+    } catch (IOException e) {
+      log.error("Error clearing cache", e);
+    }
+  }
+
+  /**
+   * Deletes cache entries that are no longer valid according to the default expiration time period.
+   */
+  public synchronized void clean() {
+    log.info("cache: cleaning");
+    try (FileLock l = lock()) {
+      deleteCacheEntries(createCleanFilter());
+    } catch (IOException e) {
+      log.error("Error cleaning cache", e);
+    }
+  }
+
+  private FileLock lock() throws IOException {
+    FileChannel ch = FileChannel.open(getLockPath(), StandardOpenOption.CREATE, StandardOpenOption.WRITE);
+    return ch.lock();
+  }
+
+  private String getKey(String uri) {
+    try {
+      MessageDigest digest = MessageDigest.getInstance(DIGEST_ALGO);
+      digest.update(uri.getBytes());
+      return byteArrayToHex(digest.digest());
+    } catch (NoSuchAlgorithmException e) {
+      throw new IllegalStateException("Couldn't create hash", e);
+    }
+  }
+
+  private void deleteCacheEntries(DirectoryStream.Filter<Path> filter) throws IOException {
+    try (DirectoryStream<Path> stream = Files.newDirectoryStream(baseDir, filter)) {
+      for (Path p : stream) {
+        try {
+          Files.delete(p);
+        } catch (Exception e) {
+          log.error("Error deleting " + p, e);
+        }
+      }
+    }
+  }
+
+  private DirectoryStream.Filter<Path> createClearFilter() throws IOException {
+    return new DirectoryStream.Filter<Path>() {
+      @Override
+      public boolean accept(Path entry) throws IOException {
+        return !LOCK_FNAME.equals(entry.getFileName().toString());
+      }
+    };
+  }
+
+  private DirectoryStream.Filter<Path> createCleanFilter() throws IOException {
+    return new DirectoryStream.Filter<Path>() {
+      @Override
+      public boolean accept(Path entry) throws IOException {
+        if (LOCK_FNAME.equals(entry.getFileName().toString())) {
+          return false;
+        }
+
+        return isCacheEntryExpired(entry, PersistentCache.this.defaultDurationToExpireMs);
+      }
+    };
+  }
+
+  private void putCache(String key, byte[] value) throws UnsupportedEncodingException, IOException {
+    Path cachePath = getCacheEntryPath(key);
+    Files.write(cachePath, value, CREATE, WRITE);
+  }
+
+  private byte[] getCache(String key) throws IOException {
+    Path cachePath = getCacheEntryPath(key);
+
+    if (!validateCacheEntry(cachePath, this.defaultDurationToExpireMs)) {
+      return null;
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Return an empty array instead of null. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1168)